### DISCUSSION TITLE
Pfsense - LTSP in VLANS (almost working)

### DISCUSSION THREAD

**Discussion by doguibnu** [ltsp/ltsp#760] [Q&A]
Hello!
How are you?

After  conversations at Pfsense forum and help, I would like share screenshot about service almost running well. . The LTSP server has 10.1.1.x IP and Vlan is 10.10.203.x. The LTSP client is trying up by VLAN, but can not finish the boot process. So follow the screenshot.

The Vlan!
![01-tela-vlan-36](https://user-images.githubusercontent.com/38897311/194856482-41c95eaf-d22d-4f03-be14-a72caa4dc129.png)

![02-tela-vlan43](https://user-images.githubusercontent.com/38897311/194856588-af98f58b-14b2-40f4-9883-b4c2c99abeaf.png)

![03-tela-screen-gate-40](https://user-images.githubusercontent.com/38897311/194856670-d5929d04-aa55-4616-bde3-8518214bc282.png)

![04-tela-tftp06-36](https://user-images.githubusercontent.com/38897311/194856732-2c36377a-6d3e-4b56-a842-b13da82fb78c.png)

The Client LTSP screenshot:

![20221007_095152](https://user-images.githubusercontent.com/38897311/194856851-47257859-76c6-4d4f-90ef-9d8e17e6c2f7.jpg)

I trying to use [PXELinux](https://ltsp.org/guides/pxelinux/) and did not works

What can I do to solve this, please!

Thank you!

Douglas

**Comment by maccadia** [comment]
Hi,

I think the issue is that LTSP has a two-step PXE boot process : 

1. Download and run `undionly.kpxe`
2. Download and run `ltsp.ipxe`

In a standard LTSP setup, dnsmasq is handling this : when the client boots, dnsmasq sets bootfile name (Default BIOS filename in Pfsense) to `undionly.kpxe` . Then, when `undionly.kpxe` runs, it does a DHCP request and dnsmasq detects it's a request from iPXE and sets bootfile name to `ltsp.ipxe`. See [ltsp-dnsmasq.conf](https://github.com/ltsp/ltsp/blob/c632102847da982703def4286f7f26f014303548/ltsp/server/dnsmasq/ltsp-dnsmasq.conf#L63).

You can see a documentation for ISC-DHCP here : https://ltsp.org/guides/isc-dhcp-server/

In my case, I worked around this by recompiling  `undionly.kpxe` from source while including this chain script : 

```bash
#!ipxe

# Chain to a custom ipxe script
# This avoid the need of using proxyDHCP (dnsmasq) or complex DHCP server configuration
# see http://ipxe.org/howto/chainloading#breaking_the_infinite_loop

ifopen
isset ${next-server} || dhcp
chain tftp://${next-server}/ltsp/ltsp.ipxe 
```

**Reply by doguibnu** [reply]
Hello! I did try but not working :(

**Reply by doguibnu** [reply]
hello!
how are you?

It does not work

**Reply by doguibnu** [reply]
Hello!

Can you explain how compile 'undionly.kpxe'?

In other topic on ipxe github an user show me link/doc from ipxe site about [looping](https://ipxe.org/howto/chainloading#breaking_the_infinite_loop)

The user answer me that undionly no have support to Vlans.
I have no experience about ipxe
But, I am stubborn to try fix and works well!

Thank you


> Hi,
> 
> I think the issue is that LTSP has a two-step PXE boot process :
> 
>     1. Download and run `undionly.kpxe`
> 
>     2. Download and run `ltsp.ipxe`
> 
> 
> In a standard LTSP setup, dnsmasq is handling this : when the client boots, dnsmasq sets bootfile name (Default BIOS filename in Pfsense) to `undionly.kpxe` . Then, when `undionly.kpxe` runs, it does a DHCP request and dnsmasq detects it's a request from iPXE and sets bootfile name to `ltsp.ipxe`. See [ltsp-dnsmasq.conf](https://github.com/ltsp/ltsp/blob/c632102847da982703def4286f7f26f014303548/ltsp/server/dnsmasq/ltsp-dnsmasq.conf#L63).
> 
> You can see a documentation for ISC-DHCP here : https://ltsp.org/guides/isc-dhcp-server/
> 
> In my case, I worked around this by recompiling `undionly.kpxe` from source while including this chain script :
> 
> ```shell
> #!ipxe
> 
> # Chain to a custom ipxe script
> # This avoid the need of using proxyDHCP (dnsmasq) or complex DHCP server configuration
> # see http://ipxe.org/howto/chainloading#breaking_the_infinite_loop
> 
> ifopen
> isset ${next-server} || dhcp
> chain tftp://${next-server}/ltsp/ltsp.ipxe 
> ```

**Reply by maccadia** [reply]
Documentation for building iPXE is here : https://ipxe.org/download#source_code

In a nutshell : 

```bash
git clone https://github.com/ipxe/ipxe.git
cd ipxe/src
# add a file named chain_script.ipxe with the appropriate content
make bin-x86_64-efi/snponly.efi bin/undionly.kpxe EMBED=chain_script.ipxe
```
Note that if you encounter an error like `lzma.h: No such file or directory` you are missing a dependency (`liblzma-dev` package for example).

**Reply by doguibnu** [reply]
Hello!
I did try run the command, but, this is the result:

![vlan-ltsp-ipxe-15-13](https://user-images.githubusercontent.com/38897311/196795346-bbad3069-4bab-4b47-861a-46dd499727be.png)

translated:
Non-existent file or directory
make[1]: *** No rule to process target 'arch/86_64/Makefile', required by 'bin-86_64-efi/deps/libgcc/__divdi3.c.d'. Stop.
make: *** [Makefile.housekeeping:270: bin-86_64-efi/snponly.efi] Error 2

apper this package: bin-86_64-efi/deps/libgcc/__divdi3.c.d  is red hat family

I think that have no solution: ipxe and pfsene no conversation!
thank you

Douglas

 



> ifopen
> isset ${next-server} || dhcp
> chain tftp://${next-server}/ltsp/ltsp.ipxe 
> ```

**Reply by maccadia** [reply]
Try it on your own computer, no need to be on the LTSP server.

**Comment by maccadia** [comment]
Looking again at your screenshots. It's weird that your `Default BIOS File Name` is set to `/srv/tftp/ltsp/undionly.kpxe`. Normally it's relative to the TFTP server's root so it should be `ltsp/undionly.kpxe`. I'm surprised that the tftp phase is working.

**Reply by doguibnu** [reply]
I get other machine to test again! I change to:  ltsp/undionly.kpxe

and now ltsp client shows me:

Pfsense:
![tela-pfsense37-37](https://user-images.githubusercontent.com/38897311/195612514-9d9f5989-5999-4b0f-96d8-36855ef1edbe.png)

Ltsp cleint
![CYMERA_20221013_101659](https://user-images.githubusercontent.com/38897311/195611141-6a8d5657-2847-47e9-9ebd-73bded9c58a0.jpg)

I am lost!

**Comment by alkisg** [comment]
Converting this issue to a discussion as it doesn't require any changes in the LTSP code base.

**Comment by doguibnu** [comment]
Now I follow again [PXELinux](https://ltsp.org/guides/pxelinux/)
And get this on LTSP client screenshot, after the ltsp client reboot alone. Stay loop

![ultima_20221013_110122](https://user-images.githubusercontent.com/38897311/195620083-94d37a41-9f6f-4966-bd15-c01389d54f96.jpg)
